docs/architecture/adr-071-crypto-v2-multi-curve.md

```diff
@@ -0,0 +1,597 @@
+message CryptoProvider {
+    string type = 1;       // Type of the crypto provider
+    string version = 2;    // Version of the crypto provider
+    map<string, bytes> config = 3;  // Configuration data with byte array values
+    google.protobuf.Any privKey = 4; // Optional if key is stored locally
+}
+```
+
+
+<b>type</b>:
+Specifies the type of the crypto provider. This field is used to identify and differentiate between various crypto provider implementations.
+
+
+<b>version</b>:
+Indicates the version of the crypto provider using semantic versioning.
+
+
+<b>configuration (map<string, bytes>)</b>:
+Contains serialized configuration data as key-value pairs, where the key is a string and the value is a byte array.
+
+<b>privKey (google.protobuf.Any)</b>:
+An optional field that can store a private key if it is managed locally.
+
+
+The [record.proto](https://github.com/cosmos/cosmos-sdk/blob/main/proto/cosmos/crypto/keyring/v1/record.proto) file will be modified to include the `CryptoProvider` message as an optional field as follows.
+
+```protobuf
+
+// record.proto
+
+message Record {
+  string name = 1;
+  google.protobuf.Any pub_key = 2;
+
+  oneof item {
+    Local local = 3;
+    Ledger ledger = 4;
+    Multi multi = 5;
+    Offline offline = 6;
+    CryptoProvider crypto_provider = 7;
+  }
+
+  message Local {
+    google.protobuf.Any priv_key = 1;
+  }
+
+  message Ledger {
+    hd.v1.BIP44Params path = 1;
+  }
+
+  message Multi {}
+
+  message Offline {}
+}
+```
+
+##### Creating and loading a `CryptoProvider`
+
+One of the main advantages of having an encapsulated entity for the defined cryptographic operations and tools is that it allows the use of different key pairs, which could be stored in various sources and/or be of different types, within the same codebase without necessitating changes to the rest of the app's code.
+
+To provide this functionality, we propose a *factory pattern* for the `CryptoProvider` interface and a *registry* for these builders.
+
+Below, we present the proposed interfaces and code snippets to illustrate the proposed architecture.
+
+```go
+// CryptoProviderFactory is a factory interface for creating CryptoProviders.
+// Must be implemented by each crypto provider.
+type CryptoProviderFactory interface {
+    CreateFromRecord(*Record) (CryptoProvider, error)
+    CreateFromConfig(ProviderMetadata) (CryptoProvider, error)
+}
+```
+
+**Note**: The mechanisms for storing and loading `records` will utilize the existing infrastructure from keyring/v1.
+
+**Example**: crypto provider factory and builder registry
+
+```go
+// crypto/v2/providerFactory.go
+
+var providerFactories map[string]CryptoProviderFactory
+
+// RegisterCryptoProviderFactory is a function that registers a CryptoProviderFactory for a CryptoProvider.
+func RegisterCryptoProviderFactory(provider CryptoProvider, factory CryptoProviderFactory) {
```

Comment:
Consider returning `id` so that `CreateCryptoProviderFromRecordOrConfig` can provide it

Reply:
I still think this is necessary